packages/failurechecker/src/failurechecker.ts

```diff
@@ -51,148 +60,208 @@ export const TimeMethods = {
   },
 };
 
+interface ReleasePullRequest {
+  number: number;
+  labels: string[];
+  updatedAt: number;
+}
+
+interface FailedRelease {
+  number: number;
+  reason: string;
+}
+
+function hasLabel(
+  pullRequest: ReleasePullRequest,
+  labelToFind: string
+): boolean {
+  return !!pullRequest.labels.find(label => label === labelToFind);
+}
+
+function buildIssueBody(failures: FailedRelease[]): string {
+  const list = failures
+    .map(failure => `* #${failure.number} - ${failure.reason}`)
+    .join('\n');
+  return `The following release PRs may have failed:\n\n${list}`;
+}
+
+/**
+ * Helper class for encapsulating the discovery of failed releases.
+ */
+class FailureChecker {
+  private octokit: Octokit;
+  private owner: string;
+  private repo: string;
+  private logger: GCFLogger;
+  constructor(
+    octokit: Octokit,
+    owner: string,
+    repo: string,
+    logger: GCFLogger = defaultLogger
+  ) {
+    this.octokit = octokit;
+    this.owner = owner;
+    this.repo = repo;
+    this.logger = logger;
+  }
+
+  /**
+   * Return a list of failed releases with the reason they failed.
+   * @param {string} terminalStateLabel The expected final status label
+   * @returns {FailedRelease[]} Failed releases with the reason they failed.
+   */
+  async findFailedReleases(
+    terminalStateLabel: string
+  ): Promise<FailedRelease[]> {
+    const failedReleases: FailedRelease[] = [];
+    // Look for releases explicitly marked failed - these are usually set by
+    // the release job
+    for await (const releasePullRequest of this.pullRequestIterator(
+      FAILED_LABEL
+    )) {
+      if (hasLabel(releasePullRequest, terminalStateLabel)) {
+        // release is marked as both failed and successful - we assume
+        // this job was retried and succeeded
+        continue;
+      }
+      this.logger.info(
+        `found failure for ${this.owner}/${this.repo} pr = ${
+          releasePullRequest.number
+        } labels = ${releasePullRequest.labels.join(',')}`
+      );
+      failedReleases.push({
+        number: releasePullRequest.number,
+        reason: 'The release job failed -- check the build log.',
+      });
+    }
+
+    // Look for in-progress releases - these usually failed to start the next
+    // step in the pipeline, or failed to report status back
```

Comment:
Question: How do we make sure failures are really failures, not a real in-progress?

Reply:
There's a check for the minimum time (3 hours). If the PR is newer than 3 hours, we ignore.

Reply:
Ah ok, i would add that in the comment

Reply:
Added to the comment.